v2/spiffeid/id_test.go

```diff
@@ -10,476 +10,440 @@ import (
-			if test.expectedError != "" {
-				require.Error(t, err)
-				assert.Contains(t, err.Error(), test.expectedError)
-				assert.Empty(t, id)
-				return
-			}
-
-			assert.NoError(t, err)
-			assert.Equal(t, test.expectedID, id)
-		})
+	// Go all the way through 255, which ensures we reject UTF-8 appropriately
+	for i := 0; i < 256; i++ {
+		if i == '/' {
+			// Don't test / since it is the delimeter between path segments
+			continue
+		}
+		s := string(rune(i))
+		suffix := fmt.Sprintf("%X", i)
+		if _, ok := tdChars[s]; ok {
+			t.Run("allow good trustdomain char "+suffix, func(t *testing.T) {
+				assertOK(t, "spiffe://trustdomain"+s+"/path", spiffeid.RequireTrustDomainFromString("trustdomain"+s), "/path")
+			})
+		} else {
+			t.Run("reject bad trustdomain char "+suffix, func(t *testing.T) {
+				assertFail(t, "spiffe://trustdomain"+s+"/path", "trust domain characters are limited to lowercase letters, numbers, dots, dashes, and underscores")
+			})
+		}
+
+		if _, ok := pathChars[s]; ok {
+			t.Run("allow good path char "+suffix, func(t *testing.T) {
+				assertOK(t, "spiffe://trustdomain/path"+s, td, "/path"+s)
+			})
+		} else {
+			t.Run("reject bad path char "+suffix, func(t *testing.T) {
+				assertFail(t, "spiffe://trustdomain/path"+s, "path segment characters are limited to letters, numbers, dots, dashes, and underscores")
+			})
+		}
 	}
-}
 
-func TestRequireFromString(t *testing.T) {
-	require.NotPanics(t, func() {
-		spiffeid.RequireFromString("spiffe://domain.test")
+	t.Run("reject bad scheme", func(t *testing.T) {
+		assertFail(t, "s", "scheme is missing or invalid")
+		assertFail(t, "spiffe:/", "scheme is missing or invalid")
+		assertFail(t, "Spiffe://", "scheme is missing or invalid")
+	})
+
+	t.Run("reject missing trust domain", func(t *testing.T) {
+		assertFail(t, "spiffe://", "trust domain is missing")
+		assertFail(t, "spiffe:///", "trust domain is missing")
+	})
+
+	t.Run("reject empty segments", func(t *testing.T) {
+		assertFail(t, "spiffe://trustdomain/", "path cannot have a trailing slash")
+		assertFail(t, "spiffe://trustdomain//", "path cannot contain empty segments")
+		assertFail(t, "spiffe://trustdomain//path", "path cannot contain empty segments")
+		assertFail(t, "spiffe://trustdomain/path/", "path cannot have a trailing slash")
+	})
+
+	t.Run("reject dot segments", func(t *testing.T) {
+		assertFail(t, "spiffe://trustdomain/.", "path cannot contain dot segments")
+		assertFail(t, "spiffe://trustdomain/./path", "path cannot contain dot segments")
+		assertFail(t, "spiffe://trustdomain/path/./other", "path cannot contain dot segments")
+		assertFail(t, "spiffe://trustdomain/path/..", "path cannot contain dot segments")
+		assertFail(t, "spiffe://trustdomain/..", "path cannot contain dot segments")
+		assertFail(t, "spiffe://trustdomain/../path", "path cannot contain dot segments")
+		assertFail(t, "spiffe://trustdomain/path/../other", "path cannot contain dot segments")
```

Comment:
We should have a test case for a valid path segment beginning with a dot
`spiffe://trustdomain/.path/other`